mat64/shadow.go

```diff
@@ -223,3 +220,55 @@ func rectanglesOverlap(off, aCols, bCols, stride int) bool {
 	// b strictly wraps and so must overlap with a.
 	return true
 }
+
+// trianglesOverlap returns whether the strided triangles a and b overlap
+// when b is offset by off elements after a but has at least one element before
+// the end of a. a and b have aSize and bSize respectively.
+func trianglesOverlap(off, aSize, bSize, stride int, aUpper, bUpper bool) bool {
+	if !rectanglesOverlap(off, aSize, bSize, stride) {
+		// Fast return if bounding rectangles do not overlap.
+		return false
+	}
+
+	if aUpper {
+		// Test overlap by projecting diagonals from the b upper
+		// left corner, and upper right or lower left corner onto
+		// the top row of a.
+		rows := off / stride
+		proj := off - rows*(stride+1)
+		if proj >= 0 && proj < aSize {
+			// Top left corner of b projects to element in top
+			// row of a.
+			return true
+		}
+		if bUpper {
+			// Project from upper right corner of b.
+			proj += bSize
+		} else {
```

Comment:
This depends whether the lower triangle is a or b. If it's a (the data with the lower address), yes because all corners may be outside of the upper triangl. If it's b, no.